galsim/phase_psf.py

```diff
@@ -0,0 +1,1017 @@
+                  "r0_500=%r, L0_inv=%r, vx=%r, vy=%r, alpha_mag=%r, rng=%r)")
+        return outstr % (self.screen_size, self.screen_scale, self.altitude, self.time_step,
+                         self.r0_500, self.L0_inv, self.vx, self.vy, self.alpha_mag, self.rng)
+
+    def __eq__(self, other):
+        return (self.screen_size == other.screen_size and
+                self.screen_scale == other.screen_scale and
+                self.altitude == other.altitude and
+                self.r0_500 == other.r0_500 and
+                self.L0_inv == other.L0_inv and
+                self.vx == other.vx and
+                self.vy == other.vy and
+                self.alpha_mag == other.alpha_mag and
+                self.rng == other.rng)
+
+    def __ne__(self, other):
+        return not self == other
+
+    def advance(self):
+        """Advance phase screen realization by self.time_step."""
+        # For ARAtmosphericScreen, since we use Fourier methods to update the screen each step
+        # anyway, it's easy to simultaneously move the screen with the wind.
+        if self.noise_frac < 1.e-12:
+            # Frozen flow (but slower than FrozenAtmosphericScreen, since using Fourier methods to
+            # move the screen in the wind).
+            self._phaseFT = self.alpha*self._phaseFT
+        else:
+            # Boiling, do a fractional random phase update.
+            self._phaseFT = self.alpha*self._phaseFT + self._noiseFT()*self.noise_frac
+        self.screen = np.fft.ifft2(self._phaseFT).real
+
+    def path_difference(self, nx, scale, theta_x=0.0*galsim.degrees, theta_y=0.0*galsim.degrees):
+        """ Compute effective pathlength differences due to phase screen.
+
+        @param nx       Size of output array
+        @param scale    Scale of output array pixels in meters
+        @param theta_x  Field angle corresponding to center of output array.
+        @param theta_y  Ditto.
+        @returns   Array of pathlength differences in nanometers.  Multiply by 2pi/wavelength to get
+                   array of phase differences.
+        """
+        img = galsim.Image(np.ascontiguousarray(self.screen), scale=self.screen_scale)
+        ii = galsim.InterpolatedImage(img, calculate_stepk=False, calculate_maxk=False,
+                                      normalization='sb')
+        ii = ii.shift(1000*self.altitude * theta_x.tan(), 1000*self.altitude * theta_y.tan())
+        return ii.drawImage(nx=nx, ny=nx, scale=scale, method='sb', dtype=np.float64).array
+
+    def reset(self):
+        """Reset phase screen back to time=0."""
+        self.rng = copy.deepcopy(self.orig_rng)
+        fx, fy = self._freq()
+        self._init_screen(fx, fy)
+
+
+class PhaseScreenList(object):
+    """ List of phase screens that can be turned into a PSF.  Screens can be either atmospheric
+    layers or optical phase screens.  Generally, one would assemble a PhaseScreenList object using
```

Comment:
You say optical phase screens here, but there don't seem to be any classes that implement that.  And as far as I can tell, the optical part is hard coded into PhaseScreenPSF.

I wouldn't be averse to refactoring the optics.py code to work in terms of phase screens as well.  This might make it easier to add in the phase screens from the optics to a PhaseScreenList in the proper way.  

Probably one screen class for the pupil size with obscuration and struts.  Then another for the aberrations.  It might be a bit of work moving things around, but I think the code to do all this is already coded up, and I think it would be worth it to put everything onto a common framework.